### ISSUE TITLE
[UX] Replace the static "A basic block for adding custom text." for custom blocks with an excerpt of the block text.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2288]
I realize that the issue title makes it sound like a feature request, but this is in fact a UX bug...

If one adds more than one custom blocks in a layout and they leave their display title blank, then these blocks get a generic "Custom block" title. The block description is a generic "A basic block for adding custom text." which is of no help in distinguishing the blocks. It get worse as more custom blocks are added to the layout:

![backdrop-custom_blocks_everywhere](https://cloud.githubusercontent.com/assets/2423362/19553734/26dafbe6-9701-11e6-92b9-dca11b5fee6b.png)

**Comment by klonos** [comment]
How about this? I randomly went with `100` for the trimmed summary length, but we can change that to whatever.

**Comment by klonos** [comment]
...much better I think:

![backdrop-issue2288-pull163-custom_block_summaries](https://cloud.githubusercontent.com/assets/2423362/19583088/85ce0dc6-9785-11e6-9613-69229ae69cf8.png)

**Comment by klonos** [comment]
...the php70 test failure seems to be an unrelated, random path URL generation failure.

**Comment by Dinornis** [comment]
Well done @klonos, this is much better.
Marking as fixed.

**Comment by serundeputy** [comment]
I'm struggling with determining if this is a bug or a feature.  Not so much because of the words, but the code is new and there is no tests coverage for it.

On the other hand it is a big UX win.

I think we can merge it and add a follow up issue to add tests for the new `getAdminPreview()` functionality.

@quicksketch and @klonos how does that sound? Obviously we need to make a decision promptly if it is to get in today 15 January 2017.

**Comment by quicksketch** [comment]
As with other UX issues, we can add this in during a bug fix release if necessary. There's another issue very similar to this one that relates to showing the custom block titles as well, though I don't have that link on hand.

**Comment by Dinornis** [comment]
Here is the link to that issue: https://github.com/backdrop/backdrop-issues/issues/1703

I like the idea of using the block content in the block preview on the layout page, but I think it should be optional, not default.

If you were to expose the Admin label field on the Add block "Custom block" form even if not required would allow users to identify the custom blocks without the need to see their content.

**Comment by quicksketch** [comment]
Thanks @Dinornis! Yeah both of these issues definitely are related.

As we're tidying up for the 1.5.3 and 1.6.0 releases and this requires a bit more testing and thinking than we can spare at the moment, I'm pushing this to 1.6.1.

**Comment by klonos** [comment]
Sorry, I filed this issue here because I forgot about #1703, but then when I tried to search for this issue, the other one came up first so I filed a PR there too. The two PRs take different approaches and I think that I like the other one better. You see, this one takes an excerpt of off the block content and that doesn't work well when the content of the block has images or links or lists.

If we leave the admin label as optional, then people are likely to be skipping it and ending up with this issue. They will then see that they "need" to use it in order to have some sanity in the "Manage block" page of layouts with many custom blocks. If we made it required, then the mess would not happen. Either way, we require additional actions from the user, while automating this by replacing the static description text with the block content plus adding the block title when/if it is available seems more intuitive and makes things work better out of the box.

Any reason why you think that having this "preview" shown by default is a bad idea @Dinornis?

**Comment by Dinornis** [comment]
I don't think this is a bad idea, just that users should be able to turn this on or off if they want to make the whole page more compact.

**Comment by klonos** [comment]
Hmm, perhaps if there is consensus we could add some sort of "live preview" checkbox at the top of each "Manage blocks" page (like we do in the Views UI) and/or a setting under `/admin/structure/layouts/settings` to have that always enabled by default.

**Comment by Dinornis** [comment]
That sounds good.

What do you think about an alternative title to this issue:
**[UX] Replace the default description of custom blocks with an excerpt of the block content.**
Also, this issue is more of an feature request than a bug.

**Comment by klonos** [comment]
> What do you think about an alternative title to this issue:

Don't mind the issue title, so long as it gets fixed 😉 

>Also, this issue is more of an feature request than a bug.

Yes, we usually have these sorts of difference of opinions with @jenlampton, but I usually end up saying something like this:

I realize that the issue seems like a feature request or a task, but what if I called it "Users cannot tell which custom block is which when they are placed in the same region because they have the same title and same description". Definitely a usability bug then. Right? 😄

**Comment by jenlampton** [comment]
If it's a user experience issue it should have [UX] in the title, but does not need the `type - bug report` label. Both tend to get a lot of attention :)

**Comment by jlfranklin** [comment]
Couldn't a Custom Block just have a title field and description?  

When the custom blocks have graphics or tons of HTML tag nonsense (usually from copy/pasting out of Word), the previews are going to either blow out the block management page (images) or be useless (tags).

Could the preview add additional CSS through style tags, and could that CSS break the block management page?

Block previews are a great idea for a contrib module.  Core's implementation should be a lighter.

**Comment by jenlampton** [comment]
I agree we shouldn't approach a preview in this issue (or in core). We can run `check_plain()` on the block body to prevent anything from getting into the description that isn't plain text.

**Comment by klonos** [comment]
This is from the [D8 Layout Initiative](https://www.drupal.org/about/strategic-initiatives/layout):

![image](https://user-images.githubusercontent.com/2423362/51070720-aac7ed00-164e-11e9-94ce-b952eff4a9c6.png)
